### PR TITLE
Add new predefined reserved roles

### PR DESCRIPTION
**Pull request by jkakavas**
This change adds two new reserved roles in Elasticsearch, `viewer`
and `editor`
Solutions will use these roles in order to provide,out of the box,
 a role for full access to data and configuration with no access to
cluster management(editor) and a role for read only access to data
again with no access to cluster management (viewer).

### CODE REVIEW

**Comment by jkakavas** [comment]
cc @bytebilly

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java [hunk 1]
```diff
@@ -1655,6 +1657,164 @@ public void testWatcherUserRole() {
...
+        assertOnlyReadAllowed(role, "kibana_sample_data_-" + randomIntBetween(0, 5));
+        assertOnlyReadAllowed(role, ".siem-signals-" + randomIntBetween(0, 5));
+        assertOnlyReadAllowed(role, "apm-" + randomIntBetween(0, 5) + "-transaction-" + randomIntBetween(0, 5));
+        assertOnlyReadAllowed(role, "logs-" + randomIntBetween(0, 5));
+        assertOnlyReadAllowed(role, "auditbeat-" + randomIntBetween(0, 5));
+        assertOnlyReadAllowed(role, "filebeat-" + randomIntBetween(0, 5));
+        assertOnlyReadAllowed(role, "packetbeat-" + randomIntBetween(0, 5));
+        assertOnlyReadAllowed(role, "winlogbeat-" + randomIntBetween(0, 5));
+        assertOnlyReadAllowed(role, "endgame-" + randomIntBetween(0, 5));
+        assertOnlyReadAllowed(role, randomAlphaOfLength(5));
```
Comment (ywangd) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java:1697]: Complement to these checks, we could use some more to ensure no access is granted for any indices begin with a dot or `ilm-history-*`.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java [hunk 2]
```diff
@@ -336,9 +336,93 @@
...
+                    "feature_actions.read",
+                    "feature_stackAlerts.read").build() },
+            null,
+            null,
+            MetadataUtils.DEFAULT_RESERVED_METADATA,
+            null);
+    }
+
+    private static RoleDescriptor buildEditorRoleDescriptor() {
+        return new RoleDescriptor("editor",
```
Comment (legrego) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:383]: @bytebilly would we expect the `editor` role to be able to create/update/delete Kibana spaces? Or is this a task reserved for administrators?

If we need the editor to manage spaces, then this role would instead need to grant the `all` application privilege, and the `reserved_ml_admin` privilege (`all` does not automatically grant ML access until 8.0):

```
RoleDescriptor.ApplicationResourcePrivileges.builder()
                    .application("kibana-.kibana")
                    .resources("*").privileges("all")
                    .build()
```

```
RoleDescriptor.ApplicationResourcePrivileges.builder()
                    .application("kibana-*")
                    .resources("*").privileges("reserved_ml_admin")
                    .build()
```

Comment (bytebilly) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:383]: @legrego I'm not sure to follow your comment. We are not using the `all` shortcut but rather assigning all individual permissions (happy to simplify if we can do differently).

Could you also explain why there is `kibana-.kibana` in the first row, and `kibana-*` in the second one? Thanks!

Comment (legrego) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:383]: Sorry for not being clear -- If we want editors to be able to manage spaces, then we have to use the `all` shortcut in order to achieve this. We don't have a dedicated "manage spaces" privilege today: https://github.com/elastic/kibana/issues/51759, and so it's currently gated behind this `all` privilege

Comment (legrego) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:383]: > Could you also explain why there is kibana-.kibana in the first row, and kibana-* in the second one? Thanks!

That is an implementation detail of Kibana's "reserved privileges", and is something we can potentially simplify in 8.x once Kibana drops support for multi-tenant setups. The `-*` at the end indicates that this privilege will be granted to all Kibana tenants, not just the default `.kibana` tenant.

Comment (bytebilly) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:383]: Larry and I had a sync on that, and he is checking if using his proposed approach will give UX problems or not. Based on that, we can define if we want to use one approach or the other.

Comment (legrego) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:383]: I didn't get a chance to verify this today. I struggled to get ES to build locally, and my time was quite limited today. I'm on PTO until Wednesday (after feature freeze), so please don't hold this PR up on me. We can tweak the application privileges definition following FF if we have to without any real interruption

**Comment by jkakavas** [comment]
ping @albertzaharovits

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java [hunk 3]
```diff
@@ -336,9 +336,60 @@
...
     }
 
+    private static RoleDescriptor buildViewerRoleDescriptor() {
+        return new RoleDescriptor(
+            "viewer",
+            new String[] {},
+            new RoleDescriptor.IndicesPrivileges[] {
+                // Stack
+                RoleDescriptor.IndicesPrivileges.builder()
+                    .indices("/~(([.]|ilm-history-).*)/")
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:351]: I thought this is supposed to look like `/@&~(([.]|ilm-history-).*)/`, but tests say it works anyway.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java [hunk 4]
```diff
@@ -336,9 +336,60 @@
...
+                    .privileges("read", "view_index_metadata").build(),
+                // Security
+                RoleDescriptor.IndicesPrivileges.builder().indices(".siem-signals-*").privileges("read", "view_index_metadata").build() },
+            new RoleDescriptor.ApplicationResourcePrivileges[] {
+                RoleDescriptor.ApplicationResourcePrivileges.builder()
+                    .application("kibana-.kibana")
+                    .resources("*")
+                    .privileges("read").build() },
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:359]: There are other indices that might be of interest such as: `.monitoring-*` and `.ml-*`, and maybe others (search for `"." in `ReservedRolesStore`).

Only pointing it out, I don't have the full picture, but those two fit the PR description "full access to data and configuration" .

**Review by albertzaharovits — APPROVED** [review]
The java code is correct 😁

**Comment by albertzaharovits** [comment]
@bytebilly  I worry about having all-encompassing hard-coded roles like this.
As soon as the user is required to think beyond "all users read/write all data" it's back to square 0, it has to learn about the low level details, ie the indices backing features used by solutions and the privileges required for them.
I think having `viewer` and `editor` per feature, eg `apm-viewer`, `ml-editor` would help with composability.

Also, since authorization is shared between Cloud/Kibana/ES, I think it would be easier for maintenance if there was such a feature that other stack components could provision roles without writing ES code, eg something like a REST role provider.

**Comment by bytebilly** [comment]
Thanks @albertzaharovits for the feedback, let me try to add more context on this iteration.

The fact that roles are "generic" is intentional, with the goal is to provide a good getting started experience that will let users to explore and share their deployments with others.
We have "per-solution" fine-grained roles (or better, building blocks) in scope for phase 2 (more [here](https://docs.google.com/document/d/1g-y6NMAa9MeR4gzq7mYsF1ik4Ye19W3ZSe4EqudzuKY/edit#heading=h.jil8adfyxnnq)).

We decided to hardcode roles for the sake of time and complexity, but I agree we could consider something more structured. We discussed to use a YML file that is "ingested" at boot time, but also the API based way is an option. Happy to discuss more on a possible proposal.

I hope this clarify, let me know if you still have concerns!

**Comment by jkakavas** [comment]
I'll merge this since the implementation captures the decided decision for phase 1 and what this should look like for the next minor version.

**Pull request merged**